### PR TITLE
refactor: create interfaces for AuthN and AuthZ

### PR DESCRIPTION
**Pull request by KeranYang**
This change defines interfaces for AuthN and AuthZ, which helps modularize the auth components in a way that each package takes care of its own business and has minimal knowledge of the implementation details of dependencies. The goal is to hide complexity and reduce engineer cognitive load.

### CODE REVIEW

File: server/authz/route_map.go [hunk 1]
```diff
@@ -20,20 +20,20 @@ package authz
...
-		RequiresAuth: requiresAuth,
+		Object:        object,
+		RequiresAuthZ: requiresAuthZ,
 	}
 }
 
-// RouteMap is a map of routes to their corresponding RouteInfo objects. This map is used to certain
-// information about the route.
+// RouteMap is a map of routes to their corresponding RouteInfo objects.
+// This map is used to save certain information about the route.
```
Comment (vigith) [inline server/authz/route_map.go:36]: lets be explicit about what information :)

**Review by vigith — APPROVED** [review]
LGTM

File: server/apis/v1/dexauthn.go [hunk 2]
```diff
@@ -98,8 +98,24 @@ func (d *DexObject) oauth2Config(scopes []string) (*oauth2.Config, error) {
 	}, nil
 }
 
-// Verify is used to validate the user ID token.
-func (d *DexObject) Verify(ctx context.Context, rawIDToken string) (*oidc.IDToken, error) {
+func (d *DexObject) Authenticate(c *gin.Context) (authn.UserIdInfo, error) {
```
Comment (whynowy) [inline server/apis/v1/dexauthn.go:101]: Call it `UserInfo`?

File: server/apis/v1/dexauthn.go [hunk 3]
```diff
@@ -98,8 +98,8 @@ func (d *DexObject) oauth2Config(scopes []string) (*oauth2.Config, error) {
 	}, nil
 }
 
-func (d *DexObject) Authenticate(c *gin.Context) (authn.UserIdInfo, error) {
-	userInfo := authn.UserIdInfo{}
+func (d *DexObject) Authenticate(c *gin.Context) (authn.UserInfo, error) {
```
Comment (whynowy) [inline server/apis/v1/dexauthn.go:101]: BTW - the common practice is returning `(pointer, error)`, which makes more sense.

Comment (KeranYang) [reply server/apis/v1/dexauthn.go]: +1.

**Review by whynowy — APPROVED** [review]

**Pull request merged**